config/environment.go

```diff
@@ -4,11 +4,114 @@
 package config
 
 import (
+	"os"
 	"reflect"
+	"strconv"
+	"strings"
 
 	"github.com/mattermost/mattermost-server/v5/model"
 )
 
+func GetEnviroment() map[string]string {
+	mmenv := make(map[string]string)
+	for _, env := range os.Environ() {
+		kv := strings.SplitN(env, "=", 2)
+		key := strings.ToUpper(kv[0])
+		if strings.HasPrefix(key, "MM") {
+			mmenv[key] = kv[1]
+		}
+	}
+
+	return mmenv
+}
+
+func applyEnvKey(key, value string, rValueSubject reflect.Value) {
+	keyParts := strings.SplitN(key, "_", 2)
+	rFieldValue := rValueSubject.FieldByNameFunc(func(candidate string) bool {
+		candidateUpper := strings.ToUpper(candidate)
+		return candidateUpper == keyParts[0]
+	})
+
+	if !rFieldValue.IsValid() {
+		return
+	}
+
+	if rFieldValue.Kind() == reflect.Ptr {
+		rFieldValue = rFieldValue.Elem()
+		if !rFieldValue.IsValid() {
+			return
+		}
+	}
+
+	switch rFieldValue.Kind() {
```

Comment:
Should we add a default case here? In case of zero values etc? Or we can check if the `reflect.Value` is assignable above.

Reply:
What would we do in the default case? Panic?

Reply:
Haven't checked but isn't there is a way to propagate error or assign a zero value?

Reply:
That's what it already does. it just gives up and assumes the env var is bad.